### PR TITLE
RavenDB-7203 Database record isn't updated after node removal

### PR DESCRIPTION
**Pull request by karmeli87**

### CODE REVIEW

File: src/Raven.Client/Server/DatabaseTopology.cs [hunk 1]
```diff
@@ -121,6 +124,7 @@ public override ulong GetTaskKey()
     
     public class DatabaseTopology
     {
+        public static bool PartOfCluster = false;
```
Comment (ayende) [inline src/Raven.Client/Server/DatabaseTopology.cs:127]: This is wrong, it shouldn't be `static`.

File: src/Raven.Server/Documents/DocumentDatabase.cs [hunk 2]
```diff
@@ -615,16 +623,48 @@ public void StateChanged(long index)
             }
         }
 
-        private void InitializeFromDatabaseRecord(DatabaseRecord dbRecord)
+        private void NotifyFeaturesAboutStateChange(DatabaseRecord record, long index)
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentDatabase.cs:626]: What about concurrency here? If we have multiple updates for the db record?

File: src/Raven.Server/Documents/DocumentDatabase.cs [hunk 3]
```diff
@@ -615,16 +623,48 @@ public void StateChanged(long index)
...
+            {
+                record = _serverStore.Cluster.ReadDatabase(context, Name,out index);
+            }
+            NotifyFeaturesAboutStateChange(record, index);
+            NotifyFeaturesAboutValueChange(record, index);
+        }
+
+        private void InitializeFromDatabaseRecord(DatabaseRecord dbRecord)
+        {
             lock (this)
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentDatabase.cs:660]: This lock isn't enough. The older version might run later, resulting in us seeing stale data.

File: src/Raven.Server/Documents/Replication/ReplicationLoader.cs [hunk 4]
```diff
@@ -402,14 +412,14 @@ private void HandleTopologyChange(DatabaseRecord newRecord, out List<OutgoingRep
             lock (_locker)
             {
                 instancesToDispose = new List<OutgoingReplicationHandler>();
-                if (newRecord == null)
+                if (newRecord == null || DatabaseTopology.PartOfCluster == false)
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/ReplicationLoader.cs:415]: Wrong, this is global, and should be local.

File: src/Raven.Server/ServerWide/ServerStore.cs [hunk 5]
```diff
@@ -343,12 +342,50 @@ public void Initialize()
...
+                DatabaseTopology.PartOfCluster = state != RachisConsensus.State.Passive;
+                if (DatabaseTopology.PartOfCluster == false)
+                {
+                    // If we are in passive state, we have prevent from tasks to be performed by this node.
+                    DisableAlloutgoingTasks();
+                }
+            }
+        }
+
+        public Task DisableAlloutgoingTasks()
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ServerStore.cs:360]: Typo

Comment (ayende) [reply src/Raven.Server/ServerWide/ServerStore.cs:360]: And why does this pass through method exists?

File: src/Raven.Server/ServerWide/ServerStore.cs [hunk 6]
```diff
@@ -343,12 +342,50 @@ public void Initialize()
...
 
-        private void OnStateChanged(object sender, RachisConsensus.State e)
+        private void OnStateChanged(object sender, RachisConsensus.State state)
         {
             using (ContextPool.AllocateOperationContext(out TransactionOperationContext context))
             using (context.OpenReadTransaction())
             {
                 NotificationCenter.Add(ClusterTopologyChanged.Create(GetClusterTopology(context), LeaderTag, NodeTag));
+                DatabaseTopology.PartOfCluster = state != RachisConsensus.State.Passive;
+                if (DatabaseTopology.PartOfCluster == false)
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ServerStore.cs:352]: Wrong, can't handle this globally.
During tests we have multiple servers, with some in cluster and some not.

File: src/Raven.Server/ServerWide/ServerStore.cs [hunk 7]
```diff
@@ -343,12 +342,50 @@ public void Initialize()
...
+                }
+            }
+        }
+
+        public Task DisableAlloutgoingTasks()
+        {
+            return DisableAlloutgoingTasksAsync();
+        }
+
+        public async Task DisableAlloutgoingTasksAsync()
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ServerStore.cs:365]: Why do it like this? We _already_ have state change, why need anything else?

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Documents/Replication/ReplicationLoader.cs (hunk 4 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/Replication/ReplicationLoader.cs:415]: Being part of a cluster is a global parameter, that affect every DB in the server.

File: src/Raven.Server/ServerWide/ServerStore.cs (hunk 6 above)
Comment (karmeli87) [reply src/Raven.Server/ServerWide/ServerStore.cs:352]: Yes, but each server is a different process, right?

Comment (ayende) [reply src/Raven.Server/ServerWide/ServerStore.cs:352]: Not in testing.

File: src/Raven.Server/Documents/Replication/ReplicationLoader.cs (hunk 4 above)
Comment (ayende) [reply src/Raven.Server/Documents/Replication/ReplicationLoader.cs:415]: You can have multiple servers in the same process. 
If you want to make this a global per server value, put in on the server instance.

**Pull request merged**